# 6.1 Data Profiling_US Educational Funding Analyis

## The script below contains the following criteria:

### 1. Import libraries
### 2. Import the dataset
### 3. Data cleaning and consistency checks
### 4. Merging states and naep dataframe
### 5. Export the clean dataset

## 1. Import libraries

In [1]:
# Importing libraries for analysis

import pandas as pd
import numpy as np
import os

## 2. Import the dataset

In [2]:
# Create path variable

path = os.getcwd()[0:os.getcwd().find('03 Scripts')]

In [3]:
path

'C:\\Users\\babum\\Desktop\\CF_Data Analytics\\Data Immersion\\Achievement 6\\04-2023 US Educational Finance Analysis\\'

In [4]:
# Import district dataframe

df_dist = pd.read_csv(os.path.join(path, '02 Data', 'Original Data','districts.csv'), index_col = False)

In [5]:
# Import naep dataframe

df_naep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data','naep.csv'), index_col = False)

In [6]:
# Import states dataframe

df_states = pd.read_csv(os.path.join(path, '02 Data', 'Original Data','states.csv'), index_col = False)

In [7]:
# Checking for the dimension of the district dataframe

df_dist.shape

(374161, 13)

In [8]:
# Checking for the dimension of the naep dataframe

df_naep.shape

(2305, 5)

In [9]:
# Checking for the dimension of the states dataframe

df_states.shape

(1275, 12)

In [10]:
# Checking the top 5 rows of the district dataframe

df_dist.head()

STATE   ENROLL                            NAME  YRDATA  TOTALREV  \
0  Alabama   9609.0  AUTAUGA COUNTY SCHOOL DISTRICT    2016     80867   
1  Alabama  30931.0  BALDWIN COUNTY SCHOOL DISTRICT    2016    338236   
2  Alabama    912.0  BARBOUR COUNTY SCHOOL DISTRICT    2016     10116   
3  Alabama   2842.0    EUFAULA CITY SCHOOL DISTRICT    2016     26182   
4  Alabama   3322.0     BIBB COUNTY SCHOOL DISTRICT    2016     32486   

   TFEDREV  TSTREV  TLOCREV  TOTALEXP  TCURINST  TCURSSVC  TCURONON  TCAPOUT  
0     7447   53842    19578     76672     43843     23941    6401.0     1506  
1    23710  145180   169346    299880    164977     97231   19439.0     9749  
2     2342    5434     2340     10070      4907      3896     975.0      110  
3     3558   15900     6724     29843     15302      7901    2274.0     4133  
4     3664   21846     6976     31662     16407     11087    3122.0      486

In [11]:
# Checking the top 5 rows of the naep dataframe

df_naep.head()

YEAR       STATE         AVG_SCORE TEST_SUBJECT  TEST_YEAR
0  2017     Alabama  232.170687741509  Mathematics          4
1  2017      Alaska  230.456277558902  Mathematics          4
2  2017     Arizona  234.435788152091  Mathematics          4
3  2017    Arkansas  233.848143678937  Mathematics          4
4  2017  California  232.262940994562  Mathematics          4

In [12]:
# Checking the top 5 rows of the states dataframe

df_states.head()

STATE  YEAR  ENROLL  TOTAL_REVENUE  FEDERAL_REVENUE  STATE_REVENUE  \
0     Alabama  1992     NaN        2678885           304177        1659028   
1      Alaska  1992     NaN        1049591           106780         720711   
2     Arizona  1992     NaN        3258079           297888        1369815   
3    Arkansas  1992     NaN        1711959           178571         958785   
4  California  1992     NaN       26260025          2072470       16546514   

   LOCAL_REVENUE  TOTAL_EXPENDITURE  INSTRUCTION_EXPENDITURE  \
0         715680            2653798                  1481703   
1         222100             972488                   498362   
2        1590376            3401580                  1435908   
3         574603            1743022                   964323   
4        7641041           27138832                 14358922   

   SUPPORT_SERVICES_EXPENDITURE  OTHER_EXPENDITURE  CAPITAL_OUTLAY_EXPENDITURE  
0                        735036                NaN                      174053  
1                        350902                NaN                       37451  
2                       1007732                NaN                      609114  
3                        483488                NaN                      145212  
4                       8520926                NaN                     2044688

## 3. Data cleaning and data consistency checks

### Data cleaning on districts dataframe

In [13]:
# Checking the datatype

df_dist.dtypes

STATE        object
ENROLL      float64
NAME         object
YRDATA        int64
TOTALREV      int64
TFEDREV       int64
TSTREV        int64
TLOCREV       int64
TOTALEXP      int64
TCURINST      int64
TCURSSVC      int64
TCURONON    float64
TCAPOUT       int64
dtype: object

In [14]:
# Check the column names

df_dist.columns

Index(['STATE', 'ENROLL', 'NAME', 'YRDATA', 'TOTALREV', 'TFEDREV', 'TSTREV',
       'TLOCREV', 'TOTALEXP', 'TCURINST', 'TCURSSVC', 'TCURONON', 'TCAPOUT'],
      dtype='object')

In [15]:
# Renaming the column name to understandable format

df_dist.rename(columns = {'STATE':'state','ENROLL':'enroll','NAME' : 'school_dist_name', 'YRDATA':'year', 'TOTALREV':'total_rev', 'TFEDREV':'fed_rev','TSTREV':'st_rev','TLOCREV':'loc_rev', 'TOTALEXP':'total_exp', 'TCURINST':'instruction_exp', 'TCURSSVC':'service_exp', 'TCURONON':'other_exp','TCAPOUT':'capital_exp'}, inplace = True)

In [16]:
df_dist.columns

Index(['state', 'enroll', 'school_dist_name', 'year', 'total_rev', 'fed_rev',
       'st_rev', 'loc_rev', 'total_exp', 'instruction_exp', 'service_exp',
       'other_exp', 'capital_exp'],
      dtype='object')

In [17]:
# Finding missing values

df_dist.isnull().sum()

state                   0
enroll              15868
school_dist_name        0
year                    0
total_rev               0
fed_rev                 0
st_rev                  0
loc_rev                 0
total_exp               0
instruction_exp         0
service_exp             0
other_exp           15868
capital_exp             0
dtype: int64

In [18]:
# Creating a new dataframe to check for the missing values

dist_nan = df_dist[df_dist['enroll'].isnull() == True]

In [19]:
dist_nan

state  enroll         school_dist_name  year  total_rev  fed_rev  \
358293  Alabama     NaN      AUTAUGA CO SCH DIST  1992      21039     2441   
358294  Alabama     NaN      BALDWIN CO SCH DIST  1992      67403     5411   
358295  Alabama     NaN      BARBOUR CO SCH DIST  1992       8554     1919   
358296  Alabama     NaN     EUFAULA CTY SCH DIST  1992      10906     1313   
358297  Alabama     NaN         BIBB CO SCH DIST  1992      11548     1838   
...         ...     ...                      ...   ...        ...      ...   
374156  Wyoming     NaN      UINTA CO SCH DIST 6  1992       6818      219   
374157  Wyoming     NaN      WASHAKIE SCH DIST 2  1992       1651       24   
374158  Wyoming     NaN   WASHAKIE CO SCH DIST 1  1992      10243      403   
374159  Wyoming     NaN     WESTON CO SCH DIST 1  1992       6645      222   
374160  Wyoming     NaN  WESTON CO SCHOOL DIST 7  1992       6304       64   

        st_rev  loc_rev  total_exp  instruction_exp  service_exp  other_exp  \
358293   14943     3655      21276            12210         5564        NaN   
358294   39272    22720      71618            34847        17269        NaN   
358295    5632     1003       8507             4738         2471        NaN   
358296    6786     2807      10066             5876         2602        NaN   
358297    8310     1400      11209             6771         3111        NaN   
...        ...      ...        ...              ...          ...        ...   
374156    4239     2360       6675             3742         2079        NaN   
374157    1050      577       1581              815          411        NaN   
374158    6514     3326      10193             6038         2696        NaN   
374159    4132     2291       6692             3881         2043        NaN   
374160    5201     1039       6754             1717          866        NaN   

        capital_exp  
358293         1544  
358294        12695  
358295          189  
358296          415  
358297          162  
...             ...  
374156          305  
374157          291  
374158          802  
374159          460  
374160         3886  

[15868 rows x 13 columns]

In [20]:
# Removing the null values by creating a new dataframe

df_dist_clean = df_dist[df_dist['enroll'].isnull() == False]

In [21]:
# Checking the output

df_dist_clean.isnull().sum()

state               0
enroll              0
school_dist_name    0
year                0
total_rev           0
fed_rev             0
st_rev              0
loc_rev             0
total_exp           0
instruction_exp     0
service_exp         0
other_exp           0
capital_exp         0
dtype: int64

#### The null values were from the year 1962. Since the data is very old and not needed for the analysis the null values are excluded by creating a new clean dataframe for districts data.

In [22]:
df_dist_clean.shape

(358293, 13)

In [23]:
# Checking for duplicates by creating a new dataframe for duplicated values

dist_dups = df_dist_clean[df_dist_clean.duplicated()]

In [24]:
dist_dups

Empty DataFrame
Columns: [state, enroll, school_dist_name, year, total_rev, fed_rev, st_rev, loc_rev, total_exp, instruction_exp, service_exp, other_exp, capital_exp]
Index: []

#### There are no duplicate values in the data.

In [25]:
# Summary statistics of the dataframe

df_dist_clean.describe()

enroll           year     total_rev       fed_rev        st_rev  \
count  3.582930e+05  358293.000000  3.582930e+05  3.582930e+05  3.582930e+05   
mean   3.134504e+03    2004.400390  3.173910e+04  2.689896e+03  1.473310e+04   
std    1.402406e+04       6.849252  1.816932e+05  1.926575e+04  8.418726e+04   
min    0.000000e+00    1993.000000  0.000000e+00  0.000000e+00  0.000000e+00   
25%    2.760000e+02    1998.000000  3.299000e+03  1.730000e+02  1.354000e+03   
50%    9.480000e+02    2004.000000  9.498000e+03  5.380000e+02  4.354000e+03   
75%    2.607000e+03    2010.000000  2.575700e+04  1.677000e+03  1.156600e+04   
max    1.077381e+06    2016.000000  2.744836e+07  3.120314e+06  1.056801e+07   

            loc_rev     total_exp  instruction_exp   service_exp  \
count  3.582930e+05  3.582930e+05     3.582930e+05  3.582930e+05   
mean   1.431610e+04  3.209838e+04     1.661785e+04  9.351062e+03   
std    8.637972e+04  1.937165e+05     1.186227e+05  4.386411e+04   
min    0.000000e+00  0.000000e+00     0.000000e+00  0.000000e+00   
25%    1.125000e+03  3.231000e+03     1.604000e+03  9.630000e+02   
50%    3.513000e+03  9.433000e+03     4.789000e+03  2.789000e+03   
75%    1.101400e+04  2.585600e+04     1.323600e+04  7.663000e+03   
max    1.514124e+07  2.962010e+07     2.076645e+07  4.786795e+06   

           other_exp   capital_exp  
count  358293.000000  3.582930e+05  
mean     1468.797724  3.157376e+03  
std      7748.388555  2.209544e+04  
min         0.000000  0.000000e+00  
25%       122.000000  9.300000e+01  
50%       380.000000  3.750000e+02  
75%      1094.000000  1.541000e+03  
max    707408.000000  3.151607e+06

### Data cleaning on naep dataframe

In [26]:
# Checking the datatype

df_naep.dtypes

YEAR             int64
STATE           object
AVG_SCORE       object
TEST_SUBJECT    object
TEST_YEAR        int64
dtype: object

In [27]:
df_naep['AVG_SCORE'] = pd.to_numeric(df_naep['AVG_SCORE'], errors = 'coerce')

In [28]:
df_naep.dtypes

YEAR              int64
STATE            object
AVG_SCORE       float64
TEST_SUBJECT     object
TEST_YEAR         int64
dtype: object

In [29]:
# Check the column names

df_naep.columns

Index(['YEAR', 'STATE', 'AVG_SCORE', 'TEST_SUBJECT', 'TEST_YEAR'], dtype='object')

In [30]:
# Renaming the column name to understandable format

df_naep.rename(columns = {'STATE':'state','YEAR':'year','AVG_SCORE' : 'avg_score', 'TEST_SUBJECT':'test_subject', 'TEST_YEAR':'test_grade'}, inplace = True)

In [31]:
df_naep.columns

Index(['year', 'state', 'avg_score', 'test_subject', 'test_grade'], dtype='object')

In [32]:
# Finding missing values

df_naep.isnull().sum()

year              0
state             0
avg_score       130
test_subject      0
test_grade        0
dtype: int64

In [33]:
# Creating a new dataframe to check for the missing values

naep_nan = df_naep[df_naep['avg_score'].isnull() == True]

In [34]:
naep_nan

year          state  avg_score test_subject  test_grade
409   2000         Alaska        NaN  Mathematics           4
413   2000       Colorado        NaN  Mathematics           4
415   2000       Delaware        NaN  Mathematics           4
417   2000        Florida        NaN  Mathematics           4
437   2000  New Hampshire        NaN  Mathematics           4
...    ...            ...        ...          ...         ...
2288  1998   North Dakota        NaN      Reading           8
2289  1998           Ohio        NaN      Reading           8
2292  1998   Pennsylvania        NaN      Reading           8
2295  1998   South Dakota        NaN      Reading           8
2299  1998        Vermont        NaN      Reading           8

[130 rows x 5 columns]

In [35]:
# Imputing the null values with 0

df_naep['avg_score'].fillna(0, inplace = True)

In [36]:
df_naep.isnull().sum()

year            0
state           0
avg_score       0
test_subject    0
test_grade      0
dtype: int64

#### The missing values in avg_score was treated by imputing 0. The missing values in this dataset is either because the reporting standards are not met or accomodations were not permitted for this assesment. Thus the values for this was imputed 0 as this can provide some information about the effect of this in funding.

In [37]:
# Checking for duplicates by creating a new dataframe for duplicated values

naep_dups = df_naep[df_naep.duplicated()]

In [38]:
naep_dups

Empty DataFrame
Columns: [year, state, avg_score, test_subject, test_grade]
Index: []

#### There are no duplicates in the dataframe.

In [39]:
# Summary statistics of the dataframe

df_naep.describe()

year    avg_score   test_grade
count  2305.000000  2305.000000  2305.000000
mean   2006.052928   234.283900     5.964425
std       7.602698    62.310452     2.000117
min    1990.000000     0.000000     4.000000
25%    2002.000000   222.695200     4.000000
50%    2007.000000   243.312240     4.000000
75%    2013.000000   268.391236     8.000000
max    2017.000000   300.568235     8.000000

In [40]:
df_naep.shape

(2305, 5)

### Data cleaning on states dataframe

In [41]:
# Checking the datatype

df_states.dtypes

STATE                            object
YEAR                              int64
ENROLL                          float64
TOTAL_REVENUE                     int64
FEDERAL_REVENUE                   int64
STATE_REVENUE                     int64
LOCAL_REVENUE                     int64
TOTAL_EXPENDITURE                 int64
INSTRUCTION_EXPENDITURE           int64
SUPPORT_SERVICES_EXPENDITURE      int64
OTHER_EXPENDITURE               float64
CAPITAL_OUTLAY_EXPENDITURE        int64
dtype: object

In [42]:
# Check the column names

df_states.columns

Index(['STATE', 'YEAR', 'ENROLL', 'TOTAL_REVENUE', 'FEDERAL_REVENUE',
       'STATE_REVENUE', 'LOCAL_REVENUE', 'TOTAL_EXPENDITURE',
       'INSTRUCTION_EXPENDITURE', 'SUPPORT_SERVICES_EXPENDITURE',
       'OTHER_EXPENDITURE', 'CAPITAL_OUTLAY_EXPENDITURE'],
      dtype='object')

In [43]:
# Renaming the column name to understandable format

df_states.rename(columns = {'STATE':'state','ENROLL':'enroll', 'YEAR':'year', 'TOTAL_REVENUE':'total_rev', 'FEDERAL_REVENUE':'fed_rev','STATE_REVENUE':'st_rev','LOCAL_REVENUE':'loc_rev', 'TOTAL_EXPENDITURE':'total_exp', 'INSTRUCTION_EXPENDITURE':'instruction_exp', 'SUPPORT_SERVICES_EXPENDITURE':'service_exp', 'OTHER_EXPENDITURE':'other_exp','CAPITAL_OUTLAY_EXPENDITURE':'capital_exp'}, inplace = True)

In [44]:
df_states.columns

Index(['state', 'year', 'enroll', 'total_rev', 'fed_rev', 'st_rev', 'loc_rev',
       'total_exp', 'instruction_exp', 'service_exp', 'other_exp',
       'capital_exp'],
      dtype='object')

In [45]:
# Finding missing values

df_states.isnull().sum()

state               0
year                0
enroll             51
total_rev           0
fed_rev             0
st_rev              0
loc_rev             0
total_exp           0
instruction_exp     0
service_exp         0
other_exp          51
capital_exp         0
dtype: int64

In [46]:
# Creating a new dataframe to check for the missing values

states_nan = df_states[df_states['enroll'].isnull() == True]

In [47]:
states_nan

state  year  enroll  total_rev  fed_rev    st_rev  \
0                Alabama  1992     NaN    2678885   304177   1659028   
1                 Alaska  1992     NaN    1049591   106780    720711   
2                Arizona  1992     NaN    3258079   297888   1369815   
3               Arkansas  1992     NaN    1711959   178571    958785   
4             California  1992     NaN   26260025  2072470  16546514   
5               Colorado  1992     NaN    3185173   163253   1307986   
6            Connecticut  1992     NaN    3834302   143542   1342539   
7               Delaware  1992     NaN     645233    45945    420942   
8   District of Columbia  1992     NaN     709480    64749         0   
9                Florida  1992     NaN   11506299   788420   5683949   
10               Georgia  1992     NaN    5536901   398701   2798674   
11                Hawaii  1992     NaN     996809    71273    903443   
12                 Idaho  1992     NaN     859329    69138    531912   
13              Illinois  1992     NaN    9748650   597077   3073178   
14               Indiana  1992     NaN    5060274   257044   2578149   
15                  Iowa  1992     NaN    2663934   125665   1176205   
16                Kansas  1992     NaN    2245348   121625    903199   
17              Kentucky  1992     NaN    2642902   289902   1643109   
18             Louisiana  1992     NaN    3341253   367822   1848734   
19                 Maine  1992     NaN    1206819    73160    511654   
20              Maryland  1992     NaN    4658498   228336   1778834   
21         Massachusetts  1992     NaN    5517659   273946   1439051   
22              Michigan  1992     NaN    9963137   537917   2539380   
23             Minnesota  1992     NaN    4684684   194092   2380754   
24           Mississippi  1992     NaN    1684706   277706    901199   
25              Missouri  1992     NaN    4043682   255194   1903124   
26               Montana  1992     NaN     828983    69947    337455   
27              Nebraska  1992     NaN    1542413    85423    492256   
28                Nevada  1992     NaN    1097076    45018    724032   
29         New Hampshire  1992     NaN    1063391    31020     86296   
30            New Jersey  1992     NaN   10814689   416063   4504718   
31            New Mexico  1992     NaN    1330091   161520    990653   
32              New York  1992     NaN   21630179  1172030   8697663   
33        North Carolina  1992     NaN    5236099   389145   3283378   
34          North Dakota  1992     NaN     559481    51476    242693   
35                  Ohio  1992     NaN    9285961   472604   3631703   
36              Oklahoma  1992     NaN    2396705   211627   1470516   
37                Oregon  1992     NaN    2773959   163544    788309   
38          Pennsylvania  1992     NaN   11257252   658139   4227323   
39          Rhode Island  1992     NaN     883073    43545    329810   
40        South Carolina  1992     NaN    2842626   249903   1350148   
41          South Dakota  1992     NaN     551668    60671    147834   
42             Tennessee  1992     NaN    3080490   308571   1305381   
43                 Texas  1992     NaN   16264180  1160185   6706959   
44                  Utah  1992     NaN    1555799   105349    872510   
45               Vermont  1992     NaN     812377    31757    204163   
46              Virginia  1992     NaN    5631293   313971   2180133   
47            Washington  1992     NaN    4870832   263053   3467779   
48         West Virginia  1992     NaN    1585099   130117   1021803   
49             Wisconsin  1992     NaN    4993909   217723   2205439   
50               Wyoming  1992     NaN     629748    31714    314216   

     loc_rev  total_exp  instruction_exp  service_exp  other_exp  capital_exp  
0     715680    2653798          1481703       735036        NaN       174053  
1     222100     972488           498362       350902        NaN        37451  
2    1590376    3401580          1

In [48]:
# Removing the null values by creating a new dataframe

states_clean = df_states[df_states['enroll'].isnull() == False]

In [49]:
# Checking the output

states_clean.isnull().sum()

state              0
year               0
enroll             0
total_rev          0
fed_rev            0
st_rev             0
loc_rev            0
total_exp          0
instruction_exp    0
service_exp        0
other_exp          0
capital_exp        0
dtype: int64

#### The null values were from the year 1962. Since the data is very old and not needed for the analysis the null values are excluded by creating a new clean dataframe for districts data.

In [50]:
# Checking for duplicates by creating a new dataframe for duplicated values

states_dups = states_clean[states_clean.duplicated()]

In [51]:
states_dups

Empty DataFrame
Columns: [state, year, enroll, total_rev, fed_rev, st_rev, loc_rev, total_exp, instruction_exp, service_exp, other_exp, capital_exp]
Index: []

#### There are no duplicates in the dataframe.

In [52]:
# Summary statistics of the dataframe

states_clean.describe()

year        enroll     total_rev       fed_rev        st_rev  \
count  1224.000000  1.224000e+03  1.224000e+03  1.224000e+03  1.224000e+03   
mean   2004.500000  9.175416e+05  9.290765e+06  7.873945e+05  4.312719e+06   
std       6.925016  1.066514e+06  1.191733e+07  1.164312e+06  5.620372e+06   
min    1993.000000  4.386600e+04  4.656500e+05  3.367200e+04  0.000000e+00   
25%    1998.750000  2.645145e+05  2.224650e+06  1.931018e+05  1.191590e+06   
50%    2004.500000  6.499335e+05  5.256748e+06  4.219105e+05  2.614030e+06   
75%    2010.250000  1.010532e+06  1.110987e+07  8.479270e+05  5.224320e+06   
max    2016.000000  6.307022e+06  8.921726e+07  9.990221e+06  5.090457e+07   

            loc_rev     total_exp  instruction_exp   service_exp  \
count  1.224000e+03  1.224000e+03     1.224000e+03  1.224000e+03   
mean   4.190651e+06  9.395936e+06     4.864428e+06  2.737271e+06   
std    5.564716e+06  1.215400e+07     6.385360e+06  3.399931e+06   
min    2.391700e+04  4.816650e+05     2.655490e+05  1.399630e+05   
25%    7.380178e+05  2.211494e+06     1.195616e+06  6.527800e+05   
50%    2.098524e+06  5.415694e+06     2.737071e+06  1.567025e+06   
75%    4.793464e+06  1.089286e+07     5.648436e+06  3.308660e+06   
max    3.610526e+07  8.532013e+07     4.396452e+07  2.605802e+07   

          other_exp   capital_exp  
count  1.224000e+03  1.224000e+03  
mean   4.299509e+05  9.242366e+05  
std    5.347893e+05  1.349417e+06  
min    1.154100e+04  1.270800e+04  
25%    1.034492e+05  1.886915e+05  
50%    2.717040e+05  5.295925e+05  
75%    5.172222e+05  9.908930e+05  
max    3.995951e+06  1.022366e+07

In [53]:
states_clean.shape

(1224, 12)

## 4. Merging the states and naep dataframe

In [54]:
# Creating a new merged dataframe

states_naep = states_clean.merge(df_naep, on = ['state', 'year'], how = 'outer', indicator = True)

In [55]:
# Checking the dimension after merging

states_naep.shape

(2917, 16)

In [56]:
# Checking for null values

states_naep.isnull().sum()

state                0
year                 0
enroll             418
total_rev          418
fed_rev            418
st_rev             418
loc_rev            418
total_exp          418
instruction_exp    418
service_exp        418
other_exp          418
capital_exp        418
avg_score          612
test_subject       612
test_grade         612
_merge               0
dtype: int64

In [57]:
# Subsetting the dataset for analysis

states_naep_clean = states_naep.loc[(states_naep['year'] >= 2005) & (states_naep['year'] <= 2016)]

In [58]:
states_naep_clean.shape

(1536, 16)

In [59]:
states_naep_clean['year'].value_counts()

2005    205
2007    205
2009    205
2011    205
2013    205
2015    205
2006     51
2008     51
2010     51
2012     51
2014     51
2016     51
Name: year, dtype: int64

#### The data is subsetted for last 10 years for analysis as other data is very old to be considered for the analysis. The naep exams are conducted every 2 years hence the count of years are 51 for the years when the test was not conducted and 205 for other years. The test is conducted on 2 subjects, reading and mathematics.

In [60]:
states_naep_clean.isnull().sum()

state                0
year                 0
enroll               6
total_rev            6
fed_rev              6
st_rev               6
loc_rev              6
total_exp            6
instruction_exp      6
service_exp          6
other_exp            6
capital_exp          6
avg_score          306
test_subject       306
test_grade         306
_merge               0
dtype: int64

In [61]:
# Creating a new dataframe to check for the missing values

states_naep_clean_nan = states_naep_clean[states_naep_clean['enroll'].isnull() == True]

In [62]:
states_naep_clean_nan

state  year  enroll  total_rev  fed_rev  st_rev  loc_rev  total_exp  \
2908  DoDEA  2015     NaN        NaN      NaN     NaN      NaN        NaN   
2909  DoDEA  2013     NaN        NaN      NaN     NaN      NaN        NaN   
2910  DoDEA  2011     NaN        NaN      NaN     NaN      NaN        NaN   
2911  DoDEA  2009     NaN        NaN      NaN     NaN      NaN        NaN   
2912  DoDEA  2007     NaN        NaN      NaN     NaN      NaN        NaN   
2913  DoDEA  2005     NaN        NaN      NaN     NaN      NaN        NaN   

      instruction_exp  service_exp  other_exp  capital_exp   avg_score  \
2908              NaN          NaN        NaN          NaN  277.191363   
2909              NaN          NaN        NaN          NaN  276.994367   
2910              NaN          NaN        NaN          NaN  272.349960   
2911              NaN          NaN        NaN          NaN  272.459316   
2912              NaN          NaN        NaN          NaN  272.692553   
2913              NaN          NaN        NaN          NaN  270.508741   

     test_subject  test_grade      _merge  
2908      Reading         8.0  right_only  
2909      Reading         8.0  right_only  
2910      Reading         8.0  right_only  
2911      Reading         8.0  right_only  
2912      Reading         8.0  right_only  
2913      Reading         8.0  right_only

In [63]:
states_merged = states_naep_clean[states_naep_clean['enroll'].isnull() == False]

In [64]:
states_merged.shape

(1530, 16)

In [65]:
states_merged.isnull().sum()

state                0
year                 0
enroll               0
total_rev            0
fed_rev              0
st_rev               0
loc_rev              0
total_exp            0
instruction_exp      0
service_exp          0
other_exp            0
capital_exp          0
avg_score          306
test_subject       306
test_grade         306
_merge               0
dtype: int64

### The missing values in merged dataframe is excluded as these are the DoDEA, Department of Defense Education Activity, data. These data are excluded from the analysis as they are very minimal data. The other null values are kept the same as those valueas are for the years when the naep test was not conducted.

## 5. Export the dataframe

In [66]:
# Export data to pkl

states_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'states_merged.pkl'))

In [67]:
# Export data to pkl

df_dist_clean.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'district_clean.pkl'))

In [68]:
states_clean.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'states_clean.pkl'))